helm/vitess/templates/_orchestrator.tpl

```diff
@@ -0,0 +1,170 @@
+###################################
+# Master Orchestrator Service
+###################################
+{{- define "orchestrator" -}}
+# set tuple values to more recognizable variables
+{{- $orc := index . 0 -}}
+
+apiVersion: v1
+kind: Service
+metadata:
+  name: orchestrator
+  labels:
+    app: vitess
+    component: orchestrator
+spec:
+  ports:
+    - name: web
+      port: 80
+      targetPort: 3000
+  selector:
+    app: vitess
+    component: orchestrator
+  type: ClusterIP
+
+---
+
+###################################
+# Headless Orchestrator Service
+###################################
+
+apiVersion: v1
+kind: Service
+metadata:
+  name: orchestrator-headless
+  annotations:
+    service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
+  labels:
+    app: vitess
+    component: orchestrator
+spec:
+  clusterIP: None
+  ports:
+    - name: web
+      port: 80
+      targetPort: 3000
+    - name: raft
+      port: 10008
+      targetPort: 10008
+  selector:
+    component: orchestrator
+    app: vitess
+
+---
+
+###################################
+# Orchestrator StatefulSet
+###################################
+apiVersion: apps/v1beta1
+kind: StatefulSet
+metadata:
+  name: orchestrator
+spec:
+  serviceName: orchestrator-headless
+  replicas: {{ $orc.replicas }}
+  podManagementPolicy: Parallel
+  updateStrategy: 
+    type: RollingUpdate
+  selector:
+    matchLabels:
+      app: vitess
+      component: orchestrator
+  template:
+    metadata:
+      labels:
+        app: vitess
+        component: orchestrator
+    spec:
+      containers:
+        - name: orchestrator
+          image: {{ $orc.image }}
+          ports:
+            - containerPort: 3000
+              name: web
+              protocol: TCP
+            - containerPort: 10008
+              name: raft
+              protocol: TCP
+          livenessProbe:
+            httpGet:
+              path: /
+              port: 3000
+            initialDelaySeconds: 300
+            timeoutSeconds: 10
+          readinessProbe:
+            httpGet:
+              path: "/api/leader-check"
+              port: 3000
+            timeoutSeconds: 10
+          volumeMounts:
+            - name: config-volume
+              mountPath: /conf/
+          resources:
+{{ toYaml ($orc.resources ) | indent 12 }}
+          env:
+            - name: MY_POD_NAME
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+            - name: VTCTLD_SERVER_PORT
+              value: "15999"
+          
+          command: ["bash"]
+          args:
+            - "-c"
+            - |
+              set -ex
+
+              # set the local config to advertise/bind its own service IP
+              sed -i -e "s/POD_NAME/$MY_POD_NAME/g" /conf/orchestrator.conf.json
```

Comment:
I'm not sure if it's ok to modify files mounted from ConfigMaps. I know if the source ConfigMap changes, the file contents will update to reflect that, so at best we're fighting with that mechanism.

We could avoid that by using an init container to mount the ConfigMap somewhere else and copy the contents to the final location (an emptyDir mounted as `/conf`). Then we can safely modify our copy, also in the init container. Then the main container can directly run the `orchestrator` binary, making the command simpler and passing SIGTERM without needing to `exec`.

Reply:
What we do here is instead of putting the file in the configmap, we put the file (as a template) in the container image. We then sed/jinja the template with values from the configmap/secrets/env/etc

Reply:
I figured that orchestrator wouldn't have a watch on the file, but I'll go ahead and use an `InitContainer` to copy it out.